trino/dbapi.py

```diff
@@ -359,12 +361,27 @@ def _format_prepared_param(self, param):
         if isinstance(param, bytes):
             return "X'%s'" % param.hex()
 
-        if isinstance(param, datetime.datetime):
-            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f %Z")
-            # strip trailing whitespace if param has no zone
-            datetime_str = datetime_str.rstrip(" ")
+        if isinstance(param, datetime.datetime) and param.tzinfo is None:
+            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f")
             return "TIMESTAMP '%s'" % datetime_str
 
+        if isinstance(param, datetime.datetime) and param.tzinfo is not None:
+            datetime_str = param.strftime("%Y-%m-%d %H:%M:%S.%f")
+            # named timezones
+            if hasattr(param.tzinfo, 'zone'):
+                return "TIMESTAMP '%s %s'" % (datetime_str, param.tzinfo.zone)
+            # offset-based timezones
+            return "TIMESTAMP '%s %s'" % (datetime_str, param.tzinfo.tzname(param))
+
+        # We can't calculate the offset for a time without a point in time
+        if isinstance(param, datetime.time) and param.tzinfo is None:
+            time_str = param.strftime("%H:%M:%S.%f")
+            return "TIME '%s'" % time_str
```

Comment:
What about time with timezone?

Reply:
All methods from tzinfo require a valid datetime to return a numeric offset. As far as I see we can't do anything more. I have added tests for both cases, but both of them are raising a NotSupportedError for now.